# Анализ поведения пользователей музыкального сервиса.

<h1>Оглавление:<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Описание-проекта:" data-toc-modified-id="Описание-проекта:-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Описание проекта:</a></span></li><li><span><a href="#Описание-данных:" data-toc-modified-id="Описание-данных:-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>Описание данных:</a></span></li></ul></li><li><span><a href="#Шаг-1.-Импортирую-необходимые-библиотеки,-открываю-файл-с-данными-и-изучаю-общую-информацию." data-toc-modified-id="Шаг-1.-Импортирую-необходимые-библиотеки,-открываю-файл-с-данными-и-изучаю-общую-информацию.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Шаг 1. Импортирую необходимые библиотеки, открываю файл с данными и изучаю общую информацию.</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Шаг-2.-Предобработка-данных." data-toc-modified-id="Шаг-2.-Предобработка-данных.-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Шаг 2. Предобработка данных.</a></span><ul class="toc-item"><li><span><a href="#Исправляю-стиль-в-заголовках-столбцов." data-toc-modified-id="Исправляю-стиль-в-заголовках-столбцов.-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Исправляю стиль в заголовках столбцов.</a></span></li><li><span><a href="#Привожу-все-значения-столбцов-со-строками-к-нижнему-регистру." data-toc-modified-id="Привожу-все-значения-столбцов-со-строками-к-нижнему-регистру.-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Привожу все значения столбцов со строками к нижнему регистру.</a></span></li><li><span><a href="#Разбираюсь-с-пропущенными-значениями." data-toc-modified-id="Разбираюсь-с-пропущенными-значениями.-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Разбираюсь с пропущенными значениями.</a></span></li><li><span><a href="#Избавляюсь-от-неявных-дубликатов-в-столбце-genre." data-toc-modified-id="Избавляюсь-от-неявных-дубликатов-в-столбце-genre.-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Избавляюсь от неявных дубликатов в столбце <code>genre</code>.</a></span></li><li><span><a href="#Удаляю-явные-дубликаты." data-toc-modified-id="Удаляю-явные-дубликаты.-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Удаляю явные дубликаты.</a></span></li><li><span><a href="#Смотрю-итоговый-результат." data-toc-modified-id="Смотрю-итоговый-результат.-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Смотрю итоговый результат.</a></span></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Шаг-3.-Проверка-гипотез." data-toc-modified-id="Шаг-3.-Проверка-гипотез.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Шаг 3. Проверка гипотез.</a></span><ul class="toc-item"><li><span><a href="#Проверяю-гипотезу-о-том,-что-активность-пользователей-зависит-от-дня-недели,-и-это-по-разному-проявляется-в-Москве-и-Санкт-Петербурге." data-toc-modified-id="Проверяю-гипотезу-о-том,-что-активность-пользователей-зависит-от-дня-недели,-и-это-по-разному-проявляется-в-Москве-и-Санкт-Петербурге.-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Проверяю гипотезу о том, что активность пользователей зависит от дня недели, и это по-разному проявляется в Москве и Санкт-Петербурге.</a></span></li><li><span><a href="#Проверяю-гипотезу-о-том,-что-в-понедельник-утром-в-Москве-преобладают-одни-жанры,-а-в-Санкт-Петербурге---другие.-Так-же-и-вечером-пятницы-преобладают-разные-жанры,-в-зависимости-от-города." data-toc-modified-id="Проверяю-гипотезу-о-том,-что-в-понедельник-утром-в-Москве-преобладают-одни-жанры,-а-в-Санкт-Петербурге---другие.-Так-же-и-вечером-пятницы-преобладают-разные-жанры,-в-зависимости-от-города.-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Проверяю гипотезу о том, что в понедельник утром в Москве преобладают одни жанры, а в Санкт-Петербурге - другие. Так же и вечером пятницы преобладают разные жанры, в зависимости от города.</a></span></li><li><span><a href="#Проверяю-гипотезу-о-том,-что-Москва-и-Санкт-Петербург-предпочитают-разные-жанры-музыки.-В-Москве-чаще-слушают-поп-музыку,-в-Санкт-Петербурге---русский-рэп." data-toc-modified-id="Проверяю-гипотезу-о-том,-что-Москва-и-Санкт-Петербург-предпочитают-разные-жанры-музыки.-В-Москве-чаще-слушают-поп-музыку,-в-Санкт-Петербурге---русский-рэп.-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Проверяю гипотезу о том, что Москва и Санкт-Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Санкт-Петербурге - русский рэп.</a></span></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></li></ul></div>

### Описание проекта:
Необходимо сравнить поведение пользователей музыкального сервиса из Москвы и Санкт-Петербурга. Цель исследования - проверка трех гипотез:
- Активность пользователей зависит от дня недели. Причём в Москве и Санкт-Петербурге это проявляется по-разному.
- В понедельник утром в Москве преобладают одни жанры, а в Санкт-Петербурге - другие. Так же и вечером пятницы преобладают разные жанры, в зависимости от города. 
- Москва и Санкт-Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Санкт-Петербурге - русский рэп.

### Описание данных:
Датасет `music` содержит:
- **`user_id`** - идентификатор пользователя;
- **`track`** - название трека;
- **`artist`** - имя исполнителя;
- **`genre`** - название жанра;
- **`city`** - город пользователя;
- **`time`** - время начала прослушивания;
- **`day`** - день недели.

## Шаг 1. Импортирую необходимые библиотеки, открываю файл с данными и изучаю общую информацию.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('/datasets/music.csv')

In [3]:
def df_info(df):
    display(df.head(3))
    print('-' * 50)
    df.info()
    print('-' * 50)
    display(df.describe(include = 'all'))
    print('-' * 50)
    print('Количество явных дубликатов:', df.duplicated().sum())

In [4]:
df_info(df)

userID                        Track            artist genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile  rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg  rock   
2    20EC38            Funiculì funiculà       Mario Lanza   pop   

             City        time        Day  
0  Saint-Petersburg  20:28:33  Wednesday  
1            Moscow  14:07:09     Friday  
2  Saint-Petersburg  20:58:07  Wednesday

--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63848 non-null  object
 2   artist    57876 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
--------------------------------------------------


userID  Track artist  genre   City        time     Day
count      65079  63848  57876  63881    65079     65079   65079
unique     41748  47245  43605    289        2     20392       3
top     A8AE9169  Intro  Sasha    pop   Moscow  08:14:07  Friday
freq          76     34      6   8850    45360        14   23149

--------------------------------------------------
Количество явных дубликатов: 3826


### Вывод:
Датасет содержит **65079** записей по **7** столбцам. Типы данных соответствуют содержимому столбцов. В названиях столбцов встречаются стилистические нарушения. В следующих столбцах содержатся пропущенные значения:
- **`track`** - **1231** пропуск;
- **`artist`** - **7203** пропуска;
- **`genre`** - **1198** пропусков.

Также датасет содержит **3826** явных дубликатов. Данные изначально плохого качества и требуют предобработки.

## Шаг 2. Предобработка данных.
### Исправляю стиль в заголовках столбцов.

In [5]:
df.columns = df.columns.str.lower()
df = df.rename(columns = {'  userid':'user_id', '  city  ':'city'})
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

### Привожу все значения столбцов со строками к нижнему регистру.
Чтобы снизить вероятность неявных дубликатов.

In [6]:
columns = ['user_id', 'track', 'artist', 'genre', 'city', 'day']
for column in columns:
    df[column] = df[column].str.lower()

### Разбираюсь с пропущенными значениями.

In [7]:
df.isna().sum()

user_id       0
track      1231
artist     7203
genre      1198
city          0
time          0
day           0
dtype: int64

Не все пропущенные значения влияют на исследование. Так в `track` и `artist` пропуски не важны для этой работы. Достаточно заменить их явными обозначениями. Но пропуски в `genre` могут помешать сравнению музыкальных вкусов в Москве и Санкт-Петербурге. Установить причину пропусков и восстановить данные не прндставляется возможным, поэтому также заполняю и эти пропуски явными обозначениями.

In [8]:
columns_to_replace = ['track', 'artist', 'genre']
for column in columns_to_replace:
    df[column] = df[column].fillna('unknown')

print(df.query('genre == "unknown" & city == "moscow"')['user_id'].count(), 'неизвестных жанра у пользователей из Москвы.')
print(df.query('genre == "unknown" & city == "saint-petersburg"')['user_id'].count(), 'неизвестных жанра у пользователей из Санкт-Петербурга.')

904 неизвестных жанра у пользователей из Москвы.
294 неизвестных жанра у пользователей из Санкт-Петербурга.


**904** неизвестных жанра у пользователей из Москвы, **294** у пользователей из Санкт-Петербурга. Позже проверю насколько это повлияло на результаты исследования.

### Избавляюсь от неявных дубликатов в столбце `genre`. 

In [9]:
df['genre'].sort_values().unique()

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'alternativepunk', 'ambient', 'americana',
       'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'author', 'avantgarde',
       'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass',
       'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo',
       'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic',
       'electropop', 'emo', 'entehno', '

Я вижу три неявных дубликата:
- hip;
- hop;
- hip-hop.

Заменяю их на hiphop.

In [10]:
def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)

In [11]:
replace_wrong_genres(['hip', 'hop', 'hip-hop'], 'hiphop')

### Удаляю явные дубликаты.

In [12]:
df = df.drop_duplicates().reset_index(drop = True)

### Смотрю итоговый результат.

In [13]:
df_info(df)

user_id                        track            artist genre  \
0  ffb692ec            kamigata to boots  the mass missile  rock   
1  55204538  delayed because of accident  andreas rönnberg  rock   
2    20ec38            funiculì funiculà       mario lanza   pop   

               city      time        day  
0  saint-petersburg  20:28:33  wednesday  
1            moscow  14:07:09     friday  
2  saint-petersburg  20:58:07  wednesday

--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61253 entries, 0 to 61252
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  61253 non-null  object
 1   track    61253 non-null  object
 2   artist   61253 non-null  object
 3   genre    61253 non-null  object
 4   city     61253 non-null  object
 5   time     61253 non-null  object
 6   day      61253 non-null  object
dtypes: object(7)
memory usage: 3.3+ MB
--------------------------------------------------


user_id    track   artist  genre    city      time     day
count      61253    61253    61253  61253   61253     61253   61253
unique     41748    46647    43567    287       2     20392       3
top     a8ae9169  unknown  unknown    pop  moscow  08:14:07  friday
freq          71     1159     6772   8323   42741        13   21840

--------------------------------------------------
Количество явных дубликатов: 0


### Вывод:
Я исправил стиль в заголовках столбцов. Привел все столбцы со строками к нижнему регистру, чтобы снизить вероятность неявных дубликатов. Заменил пропущенные значения в столбцах **`track`**, **`artist`** и **`genre`** на **unknown**. Избавился от неявных дубликатов в столбце **`genre`**. Удалил явные дубликаты.
***
Уникальных пользователей - **41748**. Самый активный пользователь - **A8AE9169**, за изучаемый период он прослушал **71** трек. Из **61253** треков, **46647** треков с уникальными названиями. Уникальных исполнителей - **43567**. Представлено **287** уникальных жанров, самый превалирующий жанр - **pop**, **8323** треков из **61253**. Сейчас самый плодотворный исполнитель **unknown**😀, а до предобработки это был - **Sasha**, представленный **6** треками. В датасете пользователи только из Москвы и Санкт-Петербурга, **70%** - **42741** пользователей из Москвы. Самое популярное время для начала прослушивания музыки - **08:14:07**, а самый популярный день недели из представленных трех - **Пятница**.

## Шаг 3. Проверка гипотез.
### Проверяю гипотезу о том, что активность пользователей зависит от дня недели, и это по-разному проявляется в Москве и Санкт-Петербурге.
Для этого просто собираю все данные в одну таблицу.

In [14]:
def cnt_tracks(day, city):
    cnt = df.query('day == @day & city == @city')['user_id'].count()
    return cnt

In [15]:
pd.DataFrame(data=[[
    'Москва', 
    cnt_tracks('monday', 'moscow'), 
    cnt_tracks('wednesday', 'moscow'), 
    cnt_tracks('friday', 'moscow'),
    df.query('city == "moscow"')['genre'].count()
], [
    'Санкт-Петербург', 
    cnt_tracks('monday', 'saint-petersburg'), 
    cnt_tracks('wednesday', 'saint-petersburg'), 
    cnt_tracks('friday', 'saint-petersburg'),
    df.query('city == "saint-petersburg"')['genre'].count()
], [
    'Оба города',
    df.query('day == "monday"')['genre'].count(),
    df.query('day == "wednesday"')['genre'].count(),
    df.query('day == "friday"')['genre'].count(),
    df['genre'].count()
]], columns=['город', 'понедельник', 'среда', 'пятница', 'всего'])

город  понедельник  среда  пятница  всего
0           Москва        15740  11056    15945  42741
1  Санкт-Петербург         5614   7003     5895  18512
2       Оба города        21354  18059    21840  61253

Разумеется в Москве и прослушиваний и пользователей больше, чем в Санкт-Петербурге. Тем не менее, данные показывают разницу в поведении пользователей:

- В Москве пик прослушиваний приходится на понедельник и пятницу, а в среду заметен спад.
- В Санкт-Петербурге, наоборот, больше слушают музыку по средам. Активность в понедельник и пятницу здесь почти в равной мере уступает среде.

**Гипотеза подтвердилась.**

### Проверяю гипотезу о том, что в понедельник утром в Москве преобладают одни жанры, а в Санкт-Петербурге - другие. Так же и вечером пятницы преобладают разные жанры, в зависимости от города.
Смотрю топ-10 жанров по городам, в понедельник утром (с 7:00 до 11:00) и в пятницу вечером (с 17:00 до 23:00).

In [16]:
moscow_general = df.query('city == "moscow"')
spb_general = df.query('city == "saint-petersburg"')

In [17]:
def genre_weekday(table, day, time1, time2):
    cnt_genre = (table.query('day == @day & @time1 < time < @time2')
                      .groupby('genre')['genre']
                      .count()
                      .sort_values(ascending = False))
    return cnt_genre.head(10)

In [18]:
genre_weekday(moscow_general, 'monday', '07:00', '11:00')

genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
Name: genre, dtype: int64

In [19]:
genre_weekday(spb_general, 'monday', '07:00', '11:00')

genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
Name: genre, dtype: int64

In [20]:
genre_weekday(moscow_general, 'friday', '17:00', '23:00')

genre
pop            713
rock           517
dance          495
electronic     482
hiphop         273
world          208
ruspop         170
alternative    163
classical      163
rusrap         142
Name: genre, dtype: int64

In [21]:
genre_weekday(spb_general, 'friday', '17:00', '23:00')

genre
pop            256
electronic     216
rock           216
dance          210
hiphop          97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
Name: genre, dtype: int64

**В понедельник, утром**:
- В Москве и Санкт-Петербурге слушают похожую музыку. Единственное отличие - в Московский топ вошёл жанр `'world'`, а в Санкт-Петербургский - `'jazz'` и `'classical'`.
- В Москве пропущенных значений оказалось так много, что значение `'unknown'` заняло десятое место среди самых популярных жанров. Значит, пропущенные значения занимают существенную долю в данных и угрожают достоверности исследования.

**В пятницу, вечером**:
- Картина не сильно изменилась. Некоторые жанры поднимаются немного выше, другие спускаются, но в целом топ-10 остаётся тем же самым.

**Гипотеза подтвердилась лишь частично.**

### Проверяю гипотезу о том, что Москва и Санкт-Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Санкт-Петербурге - русский рэп.

In [22]:
def genres(table):
    genres = table.groupby('genre')['genre'].count()
    genres = genres.sort_values(ascending = False)
    return genres.head(10)

In [23]:
genres(moscow_general)

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64

In [24]:
genres(spb_general)

genre
pop            2431
dance          1932
rock           1879
electronic     1736
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64

- Топ-6 жанров идентичен в обоих городах. Остальные жанры также совпадают, но слегка в разной последовательности.
- Поп-музыка - самый популярный жанр в Москве, как и предполагала гипотеза. Более того, в топ-10 жанров встречается близкий жанр - русская популярная музыка.
- Вопреки ожиданиям, рэп одинаково популярен в Москве и Петербурге. 

**Гипотеза подтвердилась лишь частично.**

### Общий вывод:
Гипотеза о том, что активность пользователей зависит от дня недели. Причём в Москве и Санкт-Петербурге это проявляется по-разному, **подтвердилась**. Разумеется в Москве и прослушиваний и пользователей больше, чем в Санкт-Петербурге. Тем не менее, данные показывают разницу в поведении пользователей:
- В Москве пик прослушиваний приходится на понедельник и пятницу, а в среду заметен спад.
- В Санкт-Петербурге, наоборот, больше слушают музыку по средам. Активность в понедельник и пятницу здесь почти в равной мере уступает среде.
***
Гипотеза о том, что в понедельник утром в Москве преобладают одни жанры, а в Санкт-Петербурге - другие. Так же и вечером пятницы преобладают разные жанры, в зависимости от города, **подтвердилась лишь частично**.
- В понедельник, утром:
    - В Москве и Санкт-Петербурге слушают похожую музыку. Единственное отличие - в Московский топ вошёл жанр 'world', а в Санкт-Петербургский - 'jazz' и 'classical'.
    - В Москве пропущенных значений оказалось так много, что значение 'unknown' заняло десятое место среди самых популярных жанров. Значит, пропущенные значения занимают существенную долю в данных и угрожают достоверности исследования.
- В пятницу, вечером:
    - Картина не сильно изменилась. Некоторые жанры поднимаются немного выше, другие спускаются, но в целом топ-10 остаётся тем же самым.
***
Гипотеза о том, что Москва и Санкт-Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Санкт-Петербурге - русский рэп, **подтвердилась лишь частично**.
- Топ-6 жанров идентичен в обоих городах. Остальные жанры также совпадают, но слегка в разной последовательности.
- Поп-музыка - самый популярный жанр в Москве, как и предполагала гипотеза. Более того, в топ-10 жанров встречается близкий жанр - русская популярная музыка.
- Вопреки ожиданиям, рэп одинаково популярен в Москве и Петербурге.